iotilecore/iotile/core/hw/debug/firmware_image_analyzer.py

```diff
@@ -0,0 +1,148 @@
+
+    def get_firmware_hardware_type(self):
+        """Returns hardware type specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return self.cdb_app_info['hardware_type']
+
+    def get_firmware_api_version(self):
+        """Returns firmware api version specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return (self.cdb_app_info['api_major_version'],
+                self.cdb_app_info['api_minor_version'])
+
+    def get_firmware_name(self):
+        """Returns firmware name specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return self.cdb_app_info['name'].decode("utf-8")
+
+    def get_firmware_version(self):
+        """Returns firmware module version specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return (self.cdb_app_info['module_major_version'],
+                self.cdb_app_info['module_minor_version'],
+                self.cdb_app_info['module_patch_version'])
+
+    def get_cdb_app_info(self):
+        """Gets the string of bytes of the CDB app info"""
+        cdb_app_info_bytes = self._hex_image.gets(self.max_addr - (FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO - 1), \
+                FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO)
+        if self._check_cdb_app_info(cdb_app_info_bytes):
+            self.cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+            return self.cdb_app_info
+
+        addresses = self._hex_image.addresses()
+        for address_index in range(0, self.num_addresses - 4):
+            search_bytes = self._hex_image.gets(addresses[address_index], 4)
+            search, = struct.unpack("<L", search_bytes)
+            if search == self.magic_number:
+                cdb_app_info_bytes = self._hex_image.gets(addresses[address_index - FirmwareImageAnalyzer.MAGIC_NUMBER_BYTE_OFFSET], \
+                    FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO)
+                if self._check_cdb_app_info(cdb_app_info_bytes):
+                    self.cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+                    return self.cdb_app_info
+                
+
+    def _check_cdb_app_info(self, cdb_app_info_bytes):
+        """Helper function to check the byte string if it is a valid CDB app info"""
+        _hardware_type, _api_major_version, _api_minor_version, name, _module_major_version, \
+            _module_minor_version, _module_patch_version, _num_slave_commands, _num_required_configs, \
+            _num_total_configs, _reserved, _p_config_variables, _p_slave_handlers, magic_number, \
+            _firmware_checksum = struct.unpack("<BBB6sBBBBBBBLLLL", cdb_app_info_bytes)
+        cdb_app_info = self._format_cdb_app_info(cdb_app_info_bytes)
+
+        try:
+            name = cdb_app_info['name'].decode("utf-8")
+        except:
```

Comment:
Can you comment on why this exception should be passed? Looking at the try, it isn't entirely clear why failing to decode should just return False. I'm assuming it's just to inform the user that their input wasn't valid, but in that case, should we not raise an exception to that extent and handle it in the caller?

If you choose to eat this exception, pylint will complain about it (because we usually want to raise the exception to see what went wrong further up the call chain), so you typically want to annotate it:
```suggestion
        except:  # pylint:disable=bare-except;reason we are using this anyways
```

Reply:
It also overlaps with the last return False in this function, which is returning false for a different reason -- so I would imagine that there are two different modes of `return False` that are being bundled here (the second one being that it actually doesn't pass the conditional, and the first one being that it could have failed to decode the bytes)

Reply:
@mattrunchey So some background knowledge... the CDB app info block looks like the following structure:
```
typedef struct
{
    uint8_t                 hardware_type;
    uint8_t                 api_major_version;
    uint8_t                 api_minor_version;

    char                    name[kCDBModuleNameLength];

    uint8_t                 module_major_version;
    uint8_t                 module_minor_version;        
    uint8_t                 module_patch_version;

    uint8_t                 num_slave_commands;
    uint8_t                 num_required_configs;
    uint8_t                 num_total_configs;
    
    uint8_t                 reserved;

    const cdb_config_entry  *config_variables;
    const cdb_slave_entry   *slave_handlers;        

    uint32_t                magic_number;
    uint32_t                firmware_checksum;
} CDBApplicationInfoBlock;
```
And when we build our controller, we include one of these at the end of our firmware. Some legacy controller images may have it in different locations of our firmware.

So in `get_cdb_app_info`, I first check the most common location where the app info might be -- at the end of the image. So to verify ( `_check_cdb_app_info`) that the location I'm looking at is indeed a `cdb app info`, I take a look at its magic number, along with the name. The name needs to be a string (or char array) that is of one of the recognizable firmware names. If the `name` in that location is just some random string of bytes that are not `utf-8`, it will throw an error. 

So tldr;
`_check_cdb_app_info` is supposed to verify a block of memory is a `cdb app info` block by checking the name and magic number. Ideally, it would look like 
` if cdb_app_info['name'].decode("utf-8") in KNOWN_FIRMWARE_NAMES and cdb_app_info['magic_number'] == self.magic_number`, but `cdb_app_info['name'].decode("utf-8")` can error if its not a decodeable string, thus making that memory block NOT a `cdb app info`.

Reply:
Added better exception handling.